controllers/shooter.py

```diff
@@ -1,13 +1,80 @@
+from magicbot import StateMachine, state
+
 from components.indexer import Indexer
+from components.shooter import Shooter
+from components.turret import Turret
+from components.vision import Vision
+
 
+# class ShooterController(StateMachine):
 class ShooterController:
-    """Class to control the entire shooter conglomerate"""
+    """Statemachine for high level control of the shooter and injector"""
 
     indexer: Indexer
+    shooter: Shooter
+    turret: Turret
+    vision: Vision
+
+    def __init__(self):
+        # super().__init__()
+        self.state = self.searching
+    
+    def execute(self):
+        """
+        tempoary replacement of magicbot statemachine
+        """
+        self.state()
+
+
+    # @state(first=True)
+    def searching(self):
+        """
+        The vision system does not have a target, we try to find one using odometry
+        """
+        # currently just waits for vision
+        if self.vision.get_vision_data()[2] != -1:  # -1 means no data is available
+            # print(f"searching -> tracking {self.vision.get_vision_data()}")
+            # self.next_state("tracking")
+            self.state = self.tracking
+
+    # @state
+    def tracking(self):
+        """
+        Aiming towards a vision target and spining up flywheels
+        """
+        (
+            dist,
+            delta_angle,
+            timestamp,
+        ) = self.vision.get_vision_data()  # collect data only once per loop
+        if timestamp == -1:
+            # self.next_state("searching")
+            # print(f"tracking -> searching {self.vision.get_vision_data()}")
+            self.state = self.searching
+        else:
+            self.shooter.set_range(dist)
+            self.turret.slew(delta_angle)
+            if (
```

Comment:
Consider breaking this if statement up as follows: pull all the "ready" conditions into a function, called something like "ready_to_fire", and then check if the fire command has been given. This would be clearer to read and would separate the hardware being ready from the driver's intention.